File: threading/python/10_print_in_order.py
"""
QUESTION 10: Print In Order (LeetCode 1114 - Databricks/Glean)
==============================================================

Problem Statement:
------------------
Three threads will call first(), second(), third() potentially in any order.
Ensure they print "first", "second", "third" in order regardless of 
which thread starts first.

Companies: Databricks, Glean - Tests understanding of thread ordering

Key Concepts:
- Thread ordering/sequencing
- Condition variables
- Semaphores for ordering
"""

import threading
import time


# ============================================================================
# SOLUTION 1: Using Condition Variables
# ============================================================================

class PrintInOrderCondition:
    """
    EXPLANATION:
    ------------
    Use a state variable to track progress.
    Each method waits until the previous method has completed.
    
    State transitions: 0 -> 1 -> 2 -> done
    """
    
    def __init__(self):
        self.state = 0
        self.condition = threading.Condition()
    
    def first(self, print_first):
        with self.condition:
            print_first()
            self.state = 1
            self.condition.notify_all()
    
    def second(self, print_second):
        with self.condition:
            while self.state < 1:
                self.condition.wait()
            print_second()
            self.state = 2
            self.condition.notify_all()
    
    def third(self, print_third):
        with self.condition:
            while self.state < 2:
                self.condition.wait()
            print_third()


# ============================================================================
# SOLUTION 2: Using Semaphores (More Elegant)
# ============================================================================

class PrintInOrderSemaphore:
    """
    EXPLANATION:
    ------------
    Use semaphores as gates between methods.
    - gate1: Blocks second() until first() completes
    - gate2: Blocks third() until second() completes
    
    Initial state: Both gates closed (0)
    first() opens gate1 -> second() proceeds -> opens gate2 -> third() proceeds
    """
    
    def __init__(self):
        self.gate1 = threading.Semaphore(0)  # Blocks second
        self.gate2 = threading.Semaphore(0)  # Blocks third
    
    def first(self, print_first):
        print_first()
        self.gate1.release()  # Open gate for second
    
    def second(self, print_second):
        self.gate1.acquire()  # Wait for first
        print_second()
        self.gate2.release()  # Open gate for third
    
    def third(self, print_third):
        self.gate2.acquire()  # Wait for second
        print_third()


# ============================================================================
# SOLUTION 3: Using Events (Simplest)
# ============================================================================

class PrintInOrderEvent:
    """
    EXPLANATION:
    ------------
    Events are like boolean flags with wait capability.
    Each method waits for the previous event and sets its own.
    """
    
    def __init__(self):
        self.first_done = threading.Event()
        self.second_done = threading.Event()
    
    def first(self, print_first):
        print_first()
        self.first_done.set()
    
    def second(self, print_second):
        self.first_done.wait()
        print_second()
        self.second_done.set()
    
    def third(self, print_third):
        self.second_done.wait()
        print_third()


# ============================================================================
# Test all solutions
# ============================================================================

def test_solution(solution_class):
    print(f"\nTesting {solution_class.__name__}:")
    
    obj = solution_class()
    output = []
    
    def print_first():
        output.append("first")
        print("first", end=" ")
    
    def print_second():
        output.append("second")
        print("second", end=" ")
    
    def print_third():
        output.append("third")
        print("third")
    
    # Start threads in REVERSE order to test synchronization
    t3 = threading.Thread(target=obj.third, args=(print_third,))
    t2 = threading.Thread(target=obj.second, args=(print_second,))
    t1 = threading.Thread(target=obj.first, args=(print_first,))
    
    t3.start()
    time.sleep(0.01)  # Give t3 time to start waiting
    t2.start()
    time.sleep(0.01)  # Give t2 time to start waiting
    t1.start()
    
    t1.join()
    t2.join()
    t3.join()
    
    assert output == ["first", "second", "third"], f"Wrong order: {output}"
    print("✓ Test passed!")


if __name__ == "__main__":
    test_solution(PrintInOrderCondition)
    test_solution(PrintInOrderSemaphore)
    test_solution(PrintInOrderEvent)

File: threading/cpp/10_print_in_order.cpp
/*
 * QUESTION 10: Print In Order (LeetCode 1114)
 * ===========================================
 * 
 * Problem: Three threads call first(), second(), third() in any order.
 * Ensure output is always "first", "second", "third".
 * 
 * Key Concepts: condition_variable, semaphores, atomic flags
 * 
 * Compile: g++ -std=c++20 -pthread 10_print_in_order.cpp -o print_in_order
 */

#include <iostream>
#include <thread>
#include <mutex>
#include <condition_variable>
#include <semaphore>
#include <functional>

// SOLUTION 1: Condition Variable
class Foo_CV {
    /*
     * EXPLANATION:
     * State tracks progress: 0 -> 1 -> 2 -> done
     * Each method waits for previous state.
     */
private:
    std::mutex mutex_;
    std::condition_variable cv_;
    int state_ = 0;

public:
    void first(std::function<void()> printFirst) {
        std::unique_lock<std::mutex> lock(mutex_);
        printFirst();
        state_ = 1;
        cv_.notify_all();
    }

    void second(std::function<void()> printSecond) {
        std::unique_lock<std::mutex> lock(mutex_);
        cv_.wait(lock, [this] { return state_ >= 1; });
        printSecond();
        state_ = 2;
        cv_.notify_all();
    }

    void third(std::function<void()> printThird) {
        std::unique_lock<std::mutex> lock(mutex_);
        cv_.wait(lock, [this] { return state_ >= 2; });
        printThird();
    }
};

// SOLUTION 2: Semaphores (C++20)
class Foo_Sem {
    /*
     * EXPLANATION:
     * Semaphores as gates:
     * - sem1: Blocks second until first completes
     * - sem2: Blocks third until second completes
     */
private:
    std::binary_semaphore sem1_{0};
    std::binary_semaphore sem2_{0};

public:
    void first(std::function<void()> printFirst) {
        printFirst();
        sem1_.release();
    }

    void second(std::function<void()> printSecond) {
        sem1_.acquire();
        printSecond();
        sem2_.release();
    }

    void third(std::function<void()> printThird) {
        sem2_.acquire();
        printThird();
    }
};

// SOLUTION 3: Atomic flags
class Foo_Atomic {
    /*
     * EXPLANATION:
     * Atomic flags with spin-wait.
     * Simple but wastes CPU cycles (busy waiting).
     */
private:
    std::atomic<bool> first_done_{false};
    std::atomic<bool> second_done_{false};

public:
    void first(std::function<void()> printFirst) {
        printFirst();
        first_done_.store(true, std::memory_order_release);
    }

    void second(std::function<void()> printSecond) {
        while (!first_done_.load(std::memory_order_acquire)) {
            std::this_thread::yield();
        }
        printSecond();
        second_done_.store(true, std::memory_order_release);
    }

    void third(std::function<void()> printThird) {
        while (!second_done_.load(std::memory_order_acquire)) {
            std::this_thread::yield();
        }
        printThird();
    }
};

template<typename FooClass>
void test(const std::string& name) {
    std::cout << "Testing " << name << ": ";
    
    FooClass foo;
    std::string output;
    std::mutex output_mutex;
    
    auto printFirst = [&]() {
        std::lock_guard<std::mutex> lock(output_mutex);
        output += "first";
    };
    auto printSecond = [&]() {
        std::lock_guard<std::mutex> lock(output_mutex);
        output += "second";
    };
    auto printThird = [&]() {
        std::lock_guard<std::mutex> lock(output_mutex);
        output += "third";
    };
    
    // Start in REVERSE order to test synchronization
    std::thread t3([&]() { foo.third(printThird); });
    std::this_thread::sleep_for(std::chrono::milliseconds(10));
    std::thread t2([&]() { foo.second(printSecond); });
    std::this_thread::sleep_for(std::chrono::milliseconds(10));
    std::thread t1([&]() { foo.first(printFirst); });
    
    t1.join(); t2.join(); t3.join();
    
    std::cout << output;
    std::cout << (output == "firstsecondthird" ? " ✓ PASSED" : " ✗ FAILED") << std::endl;
}

int main() {
    test<Foo_CV>("Condition Variable");
    test<Foo_Sem>("Semaphore");
    test<Foo_Atomic>("Atomic");
    return 0;
}

